tests/FSharp.Compiler.ComponentTests/EmittedIL/ForLoop/ForLoopInt32.fs.RealInternalSignatureOn.OptimizeOn.il.bsl

```diff
@@ -0,0 +1,808 @@
+                                                            int32 finish) cil managed
+    {
+      .custom instance void [FSharp.Core]Microsoft.FSharp.Core.CompilationArgumentCountsAttribute::.ctor(int32[]) = ( 01 00 02 00 00 00 01 00 00 00 01 00 00 00 00 00 ) 
+      
+      .maxstack  4
+      .locals init (uint64 V_0,
+               uint64 V_1,
+               int32 V_2)
+      IL_0000:  ldarg.0
+      IL_0001:  ldarg.1
+      IL_0002:  bge.s      IL_0009
+
+      IL_0004:  ldc.i4.0
+      IL_0005:  conv.i8
+      IL_0006:  nop
+      IL_0007:  br.s       IL_0011
+
+      IL_0009:  ldarg.0
+      IL_000a:  ldarg.1
+      IL_000b:  sub
+      IL_000c:  conv.i8
+      IL_000d:  ldc.i4.1
+      IL_000e:  conv.i8
+      IL_000f:  add
+      IL_0010:  nop
+      IL_0011:  stloc.0
+      IL_0012:  ldc.i4.0
+      IL_0013:  conv.i8
+      IL_0014:  stloc.1
+      IL_0015:  ldarg.0
+      IL_0016:  stloc.2
+      IL_0017:  br.s       IL_0029
+
+      IL_0019:  ldloc.2
+      IL_001a:  call       void assembly::f<int32>(!!0)
+      IL_001f:  nop
+      IL_0020:  ldloc.2
+      IL_0021:  ldc.i4.m1
+      IL_0022:  add
+      IL_0023:  stloc.2
+      IL_0024:  ldloc.1
+      IL_0025:  ldc.i4.1
+      IL_0026:  conv.i8
+      IL_0027:  add
+      IL_0028:  stloc.1
+      IL_0029:  ldloc.1
+      IL_002a:  ldloc.0
+      IL_002b:  blt.un.s   IL_0019
+
+      IL_002d:  ret
+    } 
+
+    .method public static void  boxedLoopVar(int32 start,
+                                             int32 finish) cil managed
+    {
+      .custom instance void [FSharp.Core]Microsoft.FSharp.Core.CompilationArgumentCountsAttribute::.ctor(int32[]) = ( 01 00 02 00 00 00 01 00 00 00 01 00 00 00 00 00 ) 
+      
+      .maxstack  4
+      .locals init (uint64 V_0,
+               uint64 V_1,
+               int32 V_2,
+               int32 V_3)
+      IL_0000:  ldarg.0
+      IL_0001:  ldarg.1
+      IL_0002:  bge.s      IL_0009
+
+      IL_0004:  ldc.i4.0
+      IL_0005:  conv.i8
+      IL_0006:  nop
+      IL_0007:  br.s       IL_0011
+
+      IL_0009:  ldarg.0
+      IL_000a:  ldarg.1
+      IL_000b:  sub
+      IL_000c:  conv.i8
+      IL_000d:  ldc.i4.1
+      IL_000e:  conv.i8
+      IL_000f:  add
+      IL_0010:  nop
+      IL_0011:  stloc.0
+      IL_0012:  ldc.i4.0
+      IL_0013:  conv.i8
+      IL_0014:  stloc.1
+      IL_0015:  ldarg.0
+      IL_0016:  stloc.2
+      IL_0017:  br.s       IL_0030
+
+      IL_0019:  ldloc.2
+      IL_001a:  stloc.3
+      IL_001b:  ldloc.3
+      IL_001c:  box        [runtime]System.Int32
+      IL_0021:  call       void assembly::forceBoxing(object)
+      IL_0026:  nop
+      IL_0027:  ldloc.2
+      IL_0028:  ldc.i4.m1
+      IL_0029:  add
+      IL_002a:  stloc.2
+      IL_002b:  ldloc.1
+      IL_002c:  ldc.i4.1
+      IL_002d:  conv.i8
+      IL_002e:  add
+      IL_002f:  stloc.1
+      IL_0030:  ldloc.1
+      IL_0031:  ldloc.0
+      IL_0032:  blt.un.s   IL_0019
+
+      IL_0034:  ret
+    } 
```

Comment:
It is unclear to me why it is falling through to the range transformation for `downto` here

https://github.com/dotnet/fsharp/blob/f41a178087f41a1300885b1f59f99e7514f9b720/tests/FSharp.Compiler.ComponentTests/EmittedIL/ForLoop/ForLoopInt32.fs#L81-L83

but not for `to`

https://github.com/dotnet/fsharp/blob/f41a178087f41a1300885b1f59f99e7514f9b720/tests/FSharp.Compiler.ComponentTests/EmittedIL/ForLoop/ForLoopInt32.fs#L44-L46

Somehow, the boxing of the loop variable must be preventing the `default 'T : int` constraint from solving the typar to `int` for `downto`, even though it is solved for `to`.

Reply:
Oh, never mind: it's actually not being solved to `int` for `to`, either.

The whole thing must just be getting turned back into a simple for-loop again later down the line, perhaps here:

https://github.com/dotnet/fsharp/blob/f41a178087f41a1300885b1f59f99e7514f9b720/src/Compiler/TypedTree/TypedTreeOps.fs#L11274